### ISSUE TITLE
Anomalous Performance in CSL-Daily SLT Task (Stage 3)

### ISSUE THREAD

**Issue by lucasnhandang** [ZechengLi19/Uni-Sign#35]
Hi, I’m currently fine-tuning for SLT task with CSL-Daily dataset, using csl_stage2_weights.pth, with a batch size of 8, gradient accumulation steps of 4, and a single RTX Pro 6000 GPU. The training configuration mirrors the implementation in this repo, with adjustments made for single GPU training. Despite this, I’m encountering strange performance where test performance worsens after each epoch, even as training loss decreases:

{"train_lr": 0.0002999999999999977, "train_loss": 4.836356440005095, "test_loss": 4.892101660481206, "test_bleu1": 47.41310224576624, "test_bleu2": 34.82573160987459, "test_bleu3": 26.027306215666027, "test_bleu4": 19.874581690633132, "test_rouge": 50.341114701454835, "epoch": 0, "n_parameters": 592.63831}
{"train_lr": 0.0002999999999999977, "train_loss": 3.9101992737728617, "test_loss": 5.0460034776616975, "test_bleu1": 45.64606655966548, "test_bleu2": 33.315351125622406, "test_bleu3": 24.87974792657306, "test_bleu4": 18.95339679463644, "test_rouge": 48.427121735248306, "epoch": 1, "n_parameters": 592.63831}
{"train_lr": 0.0002999999999999977, "train_loss": 3.511412890475729, "test_loss": 5.144124313637063, "test_bleu1": 45.83619822267763, "test_bleu2": 33.35248374465634, "test_bleu3": 24.812485581420532, "test_bleu4": 18.85659863296978, "test_rouge": 48.206142700511855, "epoch": 2, "n_parameters": 592.63831}
{"train_lr": 0.0002999999999999977, "train_loss": 3.3477205509724826, "test_loss": 5.190682580735948, "test_bleu1": 45.55788000049425, "test_bleu2": 33.04255591431852, "test_bleu3": 24.4987675200455, "test_bleu4": 18.536358582077426, "test_rouge": 47.898477838745904, "epoch": 3, "n_parameters": 592.63831}
{"train_lr": 0.0002999999999999977, "train_loss": 3.275842737218608, "test_loss": 5.213101560098154, "test_bleu1": 46.09823240940903, "test_bleu2": 33.73213475945478, "test_bleu3": 25.18156790552092, "test_bleu4": 19.208646801308443, "test_rouge": 48.801377516868406, "epoch": 4, "n_parameters": 592.63831}
{"train_lr": 0.0002999999999999977, "train_loss": 3.2312595839085785, "test_loss": 5.282922031261303, "test_bleu1": 44.31878311539202, "test_bleu2": 32.086679609128915, "test_bleu3": 23.77041557283597, "test_bleu4": 18.06549987794828, "test_rouge": 47.8392067406884, "epoch": 5, "n_parameters": 592.63831}
{"train_lr": 0.0002999999999999977, "train_loss": 3.206664871132892, "test_loss": 5.3169375578562414, "test_bleu1": 44.84965146512318, "test_bleu2": 32.35527685765278, "test_bleu3": 23.98936787057942, "test_bleu4": 18.159265837059476, "test_rouge": 47.465691106788256, "epoch": 6, "n_parameters": 592.63831}

I’m also concerned that my training configuration might not align with the original setup. I've seen in another issue that you have used 4 GPUs with DeepSpeed, while I’ve modified the code for a single GPU setup. Could this difference be contributing to the issue?

Additionally, the size of the weights file is about 1.2GB, but the checkpoint file (which I finetuned) doubles in size. This could be due to a mismatch in precision, as I’m using FP32 while the dtype of the weights for checkpoint likely used BF16. Could this be affecting the performance?

I would appreciate any insights on these issues, particularly regarding the configuration for fine-tuning with a single GPU and whether adjustments are necessary.

Thank you in advance for your help!

**Comment by ZechengLi19** [comment]
@lucasnhandang If you provide the complete hyperparameter settings, I can try to conduct further analysis for you. From my experience, using bf16 or fp32 will not lead to an excessively large performance gap. You can try the single-GPU configuration with DeepSpeed, just change  localhost:0,1,2,3  to  localhost:0 .

**Comment by lucasnhandang** [comment]
@ZechengLi19 Hi, you can check the settings in the .txt file I shared. I already extracted the training config, and the hyperparameters are consistent with those reported in the paper (except GPU + grad acc step, i modified it to adapt with my hardware):

[hyperparameters_extracted.txt](https://github.com/user-attachments/files/27448857/hyperparameters_extracted.txt)

I also retrained using DeepSpeed with a single GPU setup (using localhost:0 instead of localhost:0,1,2,3). However, the behavior was still very abnormal: the training loss decreased, while the dev/test loss kept increasing. The performance gap was still notable:

| Epoch | Train Loss | Dev Loss | Test Loss | Dev BLEU-1 | Dev BLEU-4 | Test BLEU-1 | Test BLEU-4 | Dev ROUGE | Test ROUGE |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 0 | 4.9568 | 4.7855 | 4.7734 | 48.11 | 20.23 | 48.11 | 20.23 | 52.22 | 52.22 |
| 1 | 4.0904 | 4.8440 | 4.8238 | 49.55 | 21.44 | 49.55 | 21.44 | 52.47 | 52.47 |
| 2 | 3.6721 | 4.9703 | 4.9529 | 49.39 | 20.86 | 49.39 | 20.86 | 51.59 | 51.59 |
| 3 | 3.4367 | 5.0553 | 5.0381 | 49.14 | 20.49 | 49.14 | 20.49 | 50.65 | 50.65 |
| 4 | 3.3089 | 5.0755 | 5.0637 | 50.45 | 21.44 | 50.45 | 21.44 | 52.14 | 52.14 |
| 5 | 3.2362 | 5.1051 | 5.0784 | 49.70 | 22.17 | 49.70 | 22.17 | 52.15 | 52.15 |
| 6 | 3.1904 | 5.1039 | 5.0806 | 49.87 | 22.35 | 49.87 | 22.35 | 52.58 | 52.58 |
| 7 | 3.1613 | 5.1163 | 5.0981 | 50.49 | 22.65 | 50.49 | 22.65 | 53.11 | 53.11 |
| 8 | 3.1408 | 5.1509 | 5.1241 | 50.95 | 22.97 | 50.95 | 22.97 | 53.93 | 53.93 |
| 9 | 3.1254 | 5.1575 | 5.1305 | 50.96 | 22.81 | 50.96 | 22.81 | 53.48 | 53.48 |
| 10 | 3.1146 | 5.1637 | 5.1328 | 51.16 | 23.05 | 51.16 | 23.05 | 54.09 | 54.09 |
| 11 | 3.1040 | 5.1698 | 5.1450 | 51.47 | 23.33 | 51.47 | 23.33 | 54.31 | 54.31 |
| 12 | 3.0976 | 5.1924 | 5.1572 | 50.74 | 22.80 | 50.74 | 22.80 | 53.43 | 53.43 |
| 13 | 3.0901 | 5.1931 | 5.1611 | 50.91 | 23.09 | 50.91 | 23.09 | 53.79 | 54.04 |
| 14 | 3.0855 | 5.2036 | 5.2036 | 51.56 | 23.36 | 51.56 | 23.36 | 54.28 | 54.28 |
| 15 | 3.0812 | 5.2075 | 5.2075 | 51.48 | 23.34 | 51.48 | 23.34 | 54.36 | 54.36 |
| 16 | 3.0782 | 5.2072 | 5.2072 | 51.94 | 23.83 | 51.94 | 23.83 | 54.68 | 54.68 |
| 17 | 3.0767 | 5.2103 | 5.2103 | 51.97 | 23.87 | 51.97 | 23.87 | 54.73 | 54.73 |
| 18 | 3.0748 | 5.2117 | 5.2117 | 51.73 | 23.65 | 51.73 | 23.65 | 54.60 | 54.60 |
| 19 | 3.0745 | 5.2102 | 5.2102 | 51.98 | 23.78 | 51.98 | 23.78 | 54.73 | 54.73 |

**Comment by ZechengLi19** [comment]
@lucasnhandang Actually, I haven't tried running it on a single GPU myself. You could try tuning some parameters, for example reducing `gradient_accumulation_steps` to a smaller value or adjusting the learning rate.

Also, it's worth noting that there isn't necessarily a strong linear relationship between the loss and the evaluation metrics.

**Comment by lucasnhandang** [comment]
@ZechengLi19 hi, could you share the logs as well as the training plots (train/dev loss, eval metrics,...) from the original Uni-Sign training for me to reference? I’ve tried further tuning on my side, but I still can’t reproduce the results (the BLEU-1 gap is notable), so I’m quite confused.

**Comment by ZechengLi19** [comment]
@lucasnhandang Uni-Sign produced a large number of wandb logs during training, and since quite a long time has passed, I honestly do not remember exactly which run corresponds to the final reported metrics anymore. Since you are mainly checking the metric trends, I selected one of the runs and shared it here in case it is helpful for debugging or comparison: [wandb log](https://wandb.ai/sololi/sign_clip/runs/xn03zwsz/logs?nw=nwuserlizecheng19).

In addition, I would like to remind you that the weight used for stage2 should be: [weight.pth](https://huggingface.co/ZechengLi19/Uni-Sign/blob/main/csl_stage2_weight.pth). Could you also double-check whether you are using the correct weight? Another detail is that in my original training script, `gradient_accumulation_steps` was kept as `1`.

**Comment by ZechengLi19** [comment]
It seems that the Weights & Biases (W&B) link I attempted to share for a single experiment was not successfully isolated. Although I intended to share only one specific run, it appears that the shared link still exposes the entire project workspace rather than a single experiment view. As a result, I am currently unable to properly restrict the access to an individual run’s logs via a standalone link. Therefore, I have provided screenshots of the specific experiment’s W&B dashboard as a reference for your review.


<img width="1433" height="359" alt="Image" src="https://github.com/user-attachments/assets/c3b237ac-7524-4b6c-9d67-a6e25a557c07" />

<img width="1432" height="670" alt="Image" src="https://github.com/user-attachments/assets/c12899ee-342e-45c1-a1fc-ea79b1af69ba" />
